analyzetestgroups

```diff
@@ -0,0 +1,252 @@
+        statusline = f'Submission {s}, test case {tc_id}'
+        print (' '*80, end = '\r')
+        print (statusline[:80], end = '\r')
+
+    assert timelimits is not None 
+    return timelimits, submissions
+
+def pretty_yn(value: bool) -> str:
+    # green y or red n
+     return '\033[32my\033[0m' if value else '\033[91mn\033[0m'
+
+def print_table(submissions):
+    alignto = max (len(str(sub)) for sub in submissions)
+    if alignto < len('Submission'):
+        alignto = len('Submission')
+    print("\033[01m", end = '')
+    print(f"{'Submission':{alignto}} Sample  ", end = ' ')
+    print(' '.join(f'Group {i} ' for i in range(1, num_secret_groups + 1)), end = ' ')
+    print('Pts Time  Expected\033[0m')
+    for s in submissions:
+        print (f"{s.filename:{alignto}}", end = ' ')
+        for verdict in s.verdicts:
+            print (f'{verdict:17}', end = ' ')
+        print(f"{s.points:3}", end = ' ')
+        print(f"{s.maxtime:4.2f}s", end = ' ')
+
+        # get the expected grades from the file (if exist), default to AC for AC submissions
+        expected_grades = None
+        path = problempath / 'submissions' / s.fullpath
+        with open(path, encoding = 'utf-8') as sourcefile:
+            for line in sourcefile:
+                m = expected_score_pattern.search(line)
+                if m:
+                    expected_grades = m.group("grades").split()
+                    assert len(expected_grades) == num_secret_groups
+                    break
+        if not expected_grades and s.expected_total_grade == 'AC':
+            expected_grades = ['AC'] * num_secret_groups
+
+        if expected_grades:
+            summary = [pretty_yn(expected_grades[i] == s.verdicts[i+1].grade) for i in range(num_secret_groups)]
+        else:
+            summary = ['.'] * num_secret_groups
+
+        print (''.join(summary))
+
+def check_distinguished(submissions):
+    accepting_subs = defaultdict(list)
+    for s in submissions:
+        for group, v in enumerate(s.verdicts):
+            if v.grade == 'AC':
+                accepting_subs[group].append(s)
+    ok = True
+    for i in range(1, num_secret_groups):
+        for j in range(i + 1, num_secret_groups + 1):
+            if accepting_subs[i] == accepting_subs[j]:
+                print (f'\033[91mWarning: \033[0mNo submission distinguishes test groups {i} and {j}')
+                ok = False
+    if ok:
+        print ('\033[32mOK: \033[0mAll secret test groups distinguished by some submission')
+
+problempath, inputstream = parse_args()
+timelimits, submissions = parse_log(problempath, inputstream)
+num_secret_groups = max(len(s.verdicts) for s in submissions) - 1 
+assert all (len(s.verdicts) == 1 + num_secret_groups for s in submissions)
+
+submissions.sort(key=lambda d: (-d.points, d.maxtime))
+
+print_table(submissions)
+print(f"Time limit: {timelimits[0]}s, safe: {timelimits[1]}s")
+check_distinguished(submissions)
```

Comment:
wrap this in a main function to avoid variables leaking into global scope and being possible to use by accident in inner functions

Reply:
Done in  509dacd